Provide link to download the cloud sdk

etanshaul opened the pull request:
fixes #1019

etanshaul commented:
Here are a couple examples of what this looks like. It appends the download link message to all cloud sdk selection widgets (when in warning mode):

![image](https://cloud.githubusercontent.com/assets/1735744/19485065/dc02c66a-9527-11e6-826c-ca8f5a5282cf.png)

![image](https://cloud.githubusercontent.com/assets/1735744/19485095/fd148988-9527-11e6-9677-c55525061387.png)

patflynn commented on google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkPanelTest.java:
```diff
@@ -34,11 +36,18 @@
 
-  private static final String INVALID_SDK_DIR_WARNING =
-      "No Cloud SDK was found in this directory.";
+  private static final String INVALID_SDK_DIR_WARNING = "<html>\n"
```
I'm not convinced in the value of checking the exact message text here.  isn't it good enough for the test to verify that an error is being displayed when it should be?

etanshaul replied:
I'd usually agree, but in this cause there are multiple warning messages that can be displayed in the same UI field (one for empty field, and one for invalid field) so this test checks that the right one is being shown.

patflynn replied:
got it. I wonder if we should look into building an error type in the future.

this is good for now.

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkPanel.java:
```diff
@@ -43,12 +45,19 @@
   private JPanel cloudSdkPanel;
+  private JLabel warningIcon;
+
+  private static final String CLOUD_SDK_DOWNLOAD_LINK = "https://cloud.google.com/sdk/docs/"
```
I tend to like using the URL type for URL so that any typos are caught early.

etanshaul replied:
actually, since this is just a static (hardcoded) url that we verify - and not something generated by user input for example - are you ok with leaving it as a string? I don't really see the added benefit.

joaoandremartins commented:
Can you propagate this to wherever else `appengine.cloudsdk.location.missing.message` is used?

etanshaul commented:
@joaoandremartins I think it is already. I added this code to the CloudSdkPanel class; so now anywhere there is an sdk input field this will be used.

joaoandremartins commented:
There are a couple other places:
AppEngineServerModel
CloudSdkStartupPolicy

etanshaul commented:
Do those classes have a UI for cloud sdk browsing? If so, they should be using the CloudSdkPanel.

joaoandremartins commented:
`AppEngineServerModel.checkConfiguration()` has its own local dev server UI.
CloudSdkStartupPolicy just throws up on the console.

In general, importing forms from different packages is failing due to https://youtrack.jetbrains.com/issue/IDEA-110316 / https://youtrack.jetbrains.com/issue/IDEA-162625.

etanshaul commented:
@joaoandremartins please check out the updates to `AppEngineServerModel` and `CloudSdkStartupPolicy`. I fixed / refactored the validation as we discussed; I also added the download link to these too.

joaoandremartins commented on google-cloud-tools-plugin/ultimate/src/com/google/cloud/tools/intellij/appengine/server/instance/AppEngineServerModel.java:
```diff
@@ -162,20 +165,28 @@ public void checkConfiguration() throws RuntimeConfigurationException {
-    if (CloudSdkService.getInstance().getSdkHomePath() == null) {
+    CloudSdkService cloudSdkService = CloudSdkService.getInstance();
+    if (cloudSdkService.getSdkHomePath() == null
+          || StringUtils.isEmpty(cloudSdkService.getSdkHomePath().toString())) {
```
If the home path is empty, I think we should just let validateCloudSdk() take care of it, so I'd just check for null.

etanshaul replied:
I'm checking for empty because we have 2 kinds of error messages for the user. the error message shown when validateCloudSdk fails is different.

joaoandremartins replied:
They should be different. If the SDK location is empty, is because the user set it that way. In that case, we should tell them No SDK was found in that directory, rather than an SDK directory wasn't specified.

joaoandremartins replied:
An empty SDK is invalid, and that is going to be caught when you call sdk.validateCloudSdk() shortly after.

I'm considering an empty SDK different from a null SDK. For an SDK to be empty, the user has to have made it that way (previously, it's null, hence unspecified). For me, an empty string falls under "invalid string provided by user", so the messaging should be coherent.

Also, there could be a file system where an empty string is a valid location.

joaoandremartins commented on google-cloud-tools-plugin/ultimate/src/com/google/cloud/tools/intellij/appengine/server/run/CloudSdkStartupPolicy.java:
```diff
@@ -64,15 +69,27 @@ public String getDisplayString() {
+
             if (sdkService.getSdkHomePath() == null
-                || sdkService.getSdkHomePath().toString().isEmpty()) {
+                || StringUtils.isEmpty(sdkService.getSdkHomePath().toString())) {
```
Same as above here.

joaoandremartins replied:
Also, what's the benefit of using StringUtils over just String.isEmpty()?

etanshaul replied:
StringUtils#isEmpty is null safe. but I guess since its doing a toString it doesn't really matter

joaoandremartins replied:
In that case, I'd argue the code would be a bit cleaner to revert back to String.isEmpty() (and we'd cut the dependency on that apache lib).

But, ideally, what I'd really like is to get rid of this empty check altogether, for the reasons on the above comment.

etanshaul replied:
I can change it to isEmpty since StringUtils isn't adding anything here for us

joaoandremartins commented on google-cloud-tools-plugin/ultimate/src/com/google/cloud/tools/intellij/appengine/server/run/CloudSdkStartupPolicy.java:
```diff
@@ -64,15 +69,27 @@ public String getDisplayString() {
             }
-
-            if (sdkService == null) {
+            try {
```
new line before?

joaoandremartins commented on google-cloud-tools-plugin/ultimate/src/com/google/cloud/tools/intellij/appengine/server/run/CloudSdkStartupPolicy.java:
```diff
@@ -99,11 +116,6 @@ public OSProcessHandler createProcessHandler(
             executor.run();
 
             Process devappserverProcess = executor.getProcess();
-            if (devappserverProcess == null) {
```
We shouldn't get rid of this block, because it is still possible that dev_appserver blows up for some reason and we don't want it to look nasty on the IDE.

How about we throw an ExecutionException with a generic message like "The Google App Engine local development server terminated with an error."?

etanshaul replied:
I'd prefer for shit to blow up in this case. This is truely an exception condition and it would be nice to be notified (via an error report) rather than just swallowing it with a UI message. Thoughts?

joaoandremartins replied:
The way it blows with a thrown unchecked exception or an ExecutionException is really similar AFAIK. The only difference is that you'll get a nasty visible stack trace, whereas with an ExecutionException you get a user friendly notification, with a red baloon, the report to google cloud dialog, etc.

This was also a requested fix from #932

etanshaul replied:
If that's actually the case that ExecutionException's don't get swallowed by the platform (and actually report and error) we need to fix some other code I think:

The sdk validation logic (empty sdk, invalid sdk) is also using this exception. In these cases we definitely should not be reporting errors.

etanshaul replied:
Yea exactly - this is a condition that we do not anticipate so I think this is where something went genuinely wrong and we would want to know.

etanshaul commented:
@joaoandremartins I think this pr needs your approval too since you requested changes

joaoandremartins commented:
I take it that by resolving the comment, you want to go ahead and merge things your way, so go ahead :)

etanshaul commented:
Did I resolve it? I didn't think I did anything. Either way, I'll merge it
:) Unless @patflynn wants to weight in.

On Tue, Oct 18, 2016 at 4:48 PM, João Martins <notifications@github.com>
wrote:

> I take it that by resolving the comment, you want to go ahead and merge
> things your way, so go ahead :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-intellij/pull/1034#issuecomment-254634765>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABp8QPQmPUUBQspWdMSJRrylrycTiLmGks5q1TCHgaJpZM4KZ7AH>
> .
>

patflynn commented:
nothing to add.  carry on.

On Tue, Oct 18, 2016 at 4:53 PM, Etan Shaul <notifications@github.com>
wrote:

> Did I resolve it? I didn't think I did anything. Either way, I'll merge it
> :) Unless @patflynn wants to weight in.
>
> On Tue, Oct 18, 2016 at 4:48 PM, João Martins <notifications@github.com>
> wrote:
>
> > I take it that by resolving the comment, you want to go ahead and merge
> > things your way, so go ahead :)
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/GoogleCloudPlatform/google-cloud-intellij/pull/1034#
> issuecomment-254634765>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/
> ABp8QPQmPUUBQspWdMSJRrylrycTiLmGks5q1TCHgaJpZM4KZ7AH>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-intellij/pull/1034#issuecomment-254636227>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf5HUSEtjY8gP7J1HjH-hh7GI7lhm7Nks5q1THMgaJpZM4KZ7AH>
> .
>

etanshaul commented:
cool. @joaoandremartins and I can fight it out later after work and the winner gets to merge his way :)
